Tutorial to analyze the StackOverflow developer survey dataset.

Dataset contains 64k+ responses to 60 questions(although many questions are optional). 

In [2]:
# EDA from jovian.ai

import pandas as pd

In [16]:
survey_raw_df = pd.read_csv('/Users/abeygailpanganiban/Desktop/developer_survey_2020/survey_results_public20.csv')

#Load schema file as a Pandas Series with Column as the Index and QuestionText as the Value
schema_raw = pd.read_csv('/Users/abeygailpanganiban/Desktop/developer_survey_2020/survey_results_schema20.csv', index_col='Column').QuestionText

In [4]:
pd.set_option('display.max_columns',85)
pd.set_option('display.max_rows',85)

In [17]:
# now we can use schema_raw to retrieve full question text for any column in survey_raw_df
schema_raw

Column
Respondent                      Randomized respondent ID number (not in order ...
MainBranch                      Which of the following options best describes ...
Hobbyist                                                  Do you code as a hobby?
Age                             What is your age (in years)? If you prefer not...
Age1stCode                      At what age did you write your first line of c...
CompFreq                         Is that compensation weekly, monthly, or yearly?
CompTotal                       What is your current total compensation (salar...
ConvertedComp                   Salary converted to annual USD salaries using ...
Country                                                        Where do you live?
CurrencyDesc                    Which currency do you use day-to-day? If your ...
CurrencySymbol                  Which currency do you use day-to-day? If your ...
DatabaseDesireNextYear          Which database environments have you done exte...
DatabaseW

In [5]:
survey_raw_df.head()

Respondent                                         MainBranch Hobbyist  \
0           1                     I am a developer by profession      Yes   
1           2                     I am a developer by profession       No   
2           3                        I code primarily as a hobby      Yes   
3           4                     I am a developer by profession      Yes   
4           5  I used to be a developer by profession, but no...      Yes   

    Age Age1stCode CompFreq  CompTotal  ConvertedComp             Country  \
0   NaN         13  Monthly        NaN            NaN             Germany   
1   NaN         19      NaN        NaN            NaN      United Kingdom   
2   NaN         15      NaN        NaN            NaN  Russian Federation   
3  25.0         18      NaN        NaN            NaN             Albania   
4  31.0         16      NaN        NaN            NaN       United States   

     CurrencyDesc CurrencySymbol DatabaseDesireNextYear  \
0   European Euro            EUR   Microsoft SQL Server   
1  Pound sterling            GBP                    NaN   
2             NaN            NaN                    NaN   
3    Albanian lek            ALL                    NaN   
4             NaN            NaN       MySQL;PostgreSQL   

                          DatabaseWorkedWith  \
0  Elasticsearch;Microsoft SQL Server;Oracle   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4              MySQL;PostgreSQL;Redis;SQLite   

                                             DevType  \
0  Developer, desktop or enterprise applications;...   
1            Developer, full-stack;Developer, mobile   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                           EdLevel  \
0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
1     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2                                              NaN   
3  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
4     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                          Employment  \
0  Independent contractor, freelancer, or self-em...   
1                                 Employed full-time   
2                                                NaN   
3                                                NaN   
4                                 Employed full-time   

                      Ethnicity Gender  \
0  White or of European descent    Man   
1                           NaN    NaN   
2                           NaN    NaN   
3  White or of European descent    Man   
4  White or of European descent    Man   

                                          JobFactors                 JobSat  \
0  Languages, frameworks, and other technologies ...     Slightly satisfied   
1                                                NaN      Very dissatisfied   
2                                                NaN                    NaN   
3  Flex time or a flexible schedule;Office enviro...  Slightly dissatisfied   
4                                                NaN                    NaN   

                                             JobSeek  \
0       I am not interested in new job opportunities   
1       I am not interested in new job opportunities   
2                                                NaN   
3  I’m not actively looking, but I am open to new...   
4                                                NaN   

     LanguageDesireNextYear        LanguageWorkedWith  \
0    C#;HTML/CSS;JavaScript    C#;HTML/CSS;JavaScript   
1              Python;Swift          JavaScript;Swift   
2  Objective-C;Python;Swift  Objective-C;Python;Swift   
3                       NaN                       NaN   
4           Java;Ruby;Scala         HTML/CSS;Ruby;SQL   

             MiscTechDesir

In [6]:
#Review list of columns in the dataframe

survey_raw_df.columns

Index([u'Respondent', u'MainBranch', u'Hobbyist', u'Age', u'Age1stCode',
       u'CompFreq', u'CompTotal', u'ConvertedComp', u'Country',
       u'CurrencyDesc', u'CurrencySymbol', u'DatabaseDesireNextYear',
       u'DatabaseWorkedWith', u'DevType', u'EdLevel', u'Employment',
       u'Ethnicity', u'Gender', u'JobFactors', u'JobSat', u'JobSeek',
       u'LanguageDesireNextYear', u'LanguageWorkedWith',
       u'MiscTechDesireNextYear', u'MiscTechWorkedWith',
       u'NEWCollabToolsDesireNextYear', u'NEWCollabToolsWorkedWith',
       u'NEWDevOps', u'NEWDevOpsImpt', u'NEWEdImpt', u'NEWJobHunt',
       u'NEWJobHuntResearch', u'NEWLearn', u'NEWOffTopic', u'NEWOnboardGood',
       u'NEWOtherComms', u'NEWOvertime', u'NEWPurchaseResearch',
       u'NEWPurpleLink', u'NEWSOSites', u'NEWStuck', u'OpSys', u'OrgSize',
       u'PlatformDesireNextYear', u'PlatformWorkedWith', u'PurchaseWhat',
       u'Sexuality', u'SOAccount', u'SOComm', u'SOPartFreq', u'SOVisitFreq',
       u'SurveyEase', u'SurveyLeng

## Data Prep & Cleaning

Limit analysis to following areas:
* Demographics of survey respondents and global programming community
* distribution of programming skills, exp, preferences
* employment-related info, preferences, and opinions

Select subset of columns

In [7]:
selected_columns = [
    # Demographics
    'Country',
    'Age',
    'Gender',
    'EdLevel',
    'UndergradMajor',
    # Programming experience
    'Hobbyist',
    'Age1stCode',
    'YearsCode',
    'YearsCodePro',
    'LanguageWorkedWith',
    'LanguageDesireNextYear',
    'NEWLearn',
    'NEWStuck',
    # Employment
    'Employment',
    'DevType',
    'WorkWeekHrs',
    'JobSat',
    'JobFactors',
    'NEWOvertime',
    'NEWEdImpt'
]

In [8]:
len(selected_columns)

20

In [9]:
survey_df = survey_raw_df[selected_columns].copy()

In [18]:
schema = schema_raw[selected_columns]

In [19]:
survey_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64461 entries, 0 to 64460
Data columns (total 20 columns):
Country                   64072 non-null object
Age                       45446 non-null float64
Gender                    50557 non-null object
EdLevel                   57431 non-null object
UndergradMajor            50995 non-null object
Hobbyist                  64416 non-null object
Age1stCode                57900 non-null object
YearsCode                 57684 non-null object
YearsCodePro              46349 non-null object
LanguageWorkedWith        57378 non-null object
LanguageDesireNextYear    54113 non-null object
NEWLearn                  56156 non-null object
NEWStuck                  54983 non-null object
Employment                63854 non-null object
DevType                   49370 non-null object
WorkWeekHrs               41151 non-null float64
JobSat                    45194 non-null object
JobFactors                49349 non-null object
NEWOvertime              

Most columns have data type object, either because they contain values of different types or contain empty values (NaN). It appears that every column contains some empty values since Non-Null count for every column is lower than the total number of rows(64461). We'll deal with empty values and manuually adjust data type for each column on a case by case basis.

only two columns detected as numeric. 

Let's convert some other oclumns into numeric data types while ignoring non-numeric values. 

In [20]:
survey_df['Age1stCode'] = pd.to_numeric(survey_df.Age1stCode, errors='coerce')
survey_df['YearsCode'] = pd.to_numeric(survey_df.YearsCode, errors='coerce')
survey_df['YearsCodePro'] = pd.to_numeric(survey_df.YearsCodePro, errors='coerce')

In [23]:
# View some basic statistics about numeric columns
survey_df.describe()

Age    Age1stCode     YearsCode  YearsCodePro   WorkWeekHrs
count  45446.000000  57473.000000  56784.000000  44133.000000  41151.000000
mean      30.834111     15.476572     12.782051      8.869667     40.782174
std        9.585392      5.114081      9.490657      7.759961     17.816383
min        1.000000      5.000000      1.000000      1.000000      1.000000
25%       24.000000     12.000000      6.000000      3.000000     40.000000
50%       29.000000     15.000000     10.000000      6.000000     40.000000
75%       35.000000     18.000000     17.000000     12.000000     44.000000
max      279.000000     85.000000     50.000000     50.000000    475.000000

Problem with age column:
* min = 1, max = 100+. Common issue with surveys: responses may contain invalid values due to accidental or intentional errors while responding

Simple Fix: 
* ignore rows where age is higher than 100 years and lower than 10 years

In [24]:
survey_df.drop(survey_df[survey_df.Age < 10].index, inplace=True)
survey_df.drop(survey_df[survey_df.Age > 100].index, inplace=True)

Same problem with WorkWeekHrs. 
* Ignore entries where value is higher than 140 hours (~20 hrs/day)

In [25]:
survey_df.drop(survey_df[survey_df.WorkWeekHrs > 140].index, inplace=True)

Gender Column:
* column allows for picking multiple optinos. We'll remove values containing more than one option to simplify analysis.

In [26]:
survey_df['Gender'].value_counts()

Man                                                            45895
Woman                                                           3835
Non-binary, genderqueer, or gender non-conforming                385
Man;Non-binary, genderqueer, or gender non-conforming            121
Woman;Non-binary, genderqueer, or gender non-conforming           92
Woman;Man                                                         73
Woman;Man;Non-binary, genderqueer, or gender non-conforming       25
Name: Gender, dtype: int64

In [27]:
import numpy as np

In [28]:
survey_df.where(~(survey_df.Gender.str.contains(';',na=False)), np.nan, inplace=True)

Dataset prepared for analysis. Look at sample of rows from dataframe

In [29]:
survey_df.sample(10)

Country   Age Gender  \
17304                Brazil  18.0    Man   
48810         United States   NaN    NaN   
27631               Germany   NaN    Man   
35824                  Iran   NaN    NaN   
54052  United Arab Emirates   NaN    NaN   
25268         United States  28.0    Man   
19891         United States   NaN    Man   
47604           New Zealand   NaN  Woman   
18633         United States   NaN    Man   
57620               Nigeria  49.0    Man   

                                                 EdLevel  \
17304  Some college/university study without earning ...   
48810       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
27631                                                NaN   
35824       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
54052       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
25268  Some college/university study without earning ...   
19891       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
47604       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
18633  Secondary school (e.g. American high school, G...   
57620                                                NaN   

                                          UndergradMajor Hobbyist  Age1stCode  \
17304  Information systems, information technology, o...      Yes        15.0   
48810                          Mathematics or statistics      Yes         8.0   
27631  Computer science, computer engineering, or sof...      Yes        10.0   
35824  Computer science, computer engineering, or sof...       No        19.0   
54052  Information systems, information technology, o...      Yes        19.0   
25268  Computer science, computer engineering, or sof...      Yes        12.0   
19891  Another engineering discipline (such as civil,...       No        19.0   
47604                          Mathematics or statistics      Yes         8.0   
18633                                                NaN      Yes        14.0   
57620  Another engineering discipline (such as civil,...       No        45.0   

       YearsCode  YearsCodePro  \
17304        3.0           NaN   
48810       13.0           NaN   
27631        5.0           NaN   
35824       10.0           5.0   
54052        6.0           2.0   
25268       16.0           8.0   
19891       16.0          14.0   
47604       20.0          11.0   
18633        5.0           NaN   
57620        3.0           NaN   

                                      LanguageWorkedWith  \
17304                     HTML/CSS;JavaScript;TypeScript   
48810  Bash/Shell/PowerShell;C;C++;Go;Haskell;HTML/CS...   
27631                     C;C++;HTML/CSS;Java;JavaScript   
35824                              C#;JavaScript;PHP;SQL   
54052                    HTML/CSS;Java;JavaScript;Kotlin   
25268  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;K...   
19891                                        Java;Python   
47604                                   Python;R;SQL;VBA   
18633                                    C;HTML/CSS;Java   
57620                C#;HTML/CSS;Java;JavaScript;PHP;SQL   

                                  LanguageDesireNextYear  \
17304                Dart;HTML/CSS;JavaScript;TypeScript   
48810                                            Haskell   
27631                                     C;C++;HTML/CSS   
35824                                  Dart;Python;Swift   
54052                                                NaN   
25268  Bash/Shell/PowerShell;C#;JavaScript;Kotlin;Pyt...   
19891                                             Kotlin   
47604                         C#;JavaScript;Python;R;SQL   
18633                                    Dart;JavaScript   
57620                                           Assembly   

                   NEWLearn  \
17304      Every few months   
48810      Every few months   
27631  Once every few years   
35824      Every few months   
54052      Every few months   
25268      Every few months   
19891      Every few months   
47604           Once a year   
18633 